x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -154,13 +191,24 @@ impl<'a, K: Key> State<'a, K> {
             value: Vec<u8>,
         }
 
-        let mut cache = self.cache.borrow_mut();
-
-        let args: Vec<_> = cache
+        let (mut puts, mut deletes) = (Vec::new(), Vec::new());
+        self.cache
+            .borrow_mut()
             .drain()
-            .map(|(key, value)| PutArgs { key, value })
-            .collect();
-        let serialized_args = borsh::to_vec(&args).expect("failed to serialize");
-        unsafe { put_many_bytes(serialized_args.as_ptr(), serialized_args.len()) };
+            .for_each(|(key, value)| match value {
+                CachedData::Data(value) => puts.push(PutArgs { key, value }),
+                CachedData::ToDelete => deletes.push(key),
+                CachedData::Deleted => (),
+            });
+
+        if !puts.is_empty() {
+            let serialized_args = borsh::to_vec(&puts).expect("failed to serialize");
+            unsafe { put_many_bytes(serialized_args.as_ptr(), serialized_args.len()) };
+        }
+
+        if !deletes.is_empty() {
+            let serialized_args = borsh::to_vec(&deletes.as_slice()).expect("failed to serialize");
+            unsafe { delete_many_bytes(serialized_args.as_ptr(), serialized_args.len()) };
+        }
```

Comment:
Haha, I just keep seeing good reason to centralize the API into a single call. If there's only one call, you can do a simple `if cache.is_empty() { return }` for `Drop`